src/frontend/src/common/utils/locationUtils.ts

```diff
@@ -151,3 +151,70 @@ export const getIdFromCollectionLocation = (
     return collectionLocation.id;
   }
 };
+
+function findMaxDepthLocation(
+  searchLocation: GisaidLocation | NamedGisaidLocation,
+  maxDepth: keyof GisaidLocation,
+  locations: NamedGisaidLocation[],
+  locationFinderCache: LocationFinderCache | null = null
+): NamedGisaidLocation | undefined {
+  // Safety check -- Should not occur, but if app hasn't finished loading the
+  // Locations data and this gets called, would blow up because we assume there
+  // are some Locations below. Instead, fallback to `undefined` location.
+  if (locations.length === 0) {
+    return undefined;
+  }
+
+  // Only do cache interactions if one was provided
+  if (
+    locationFinderCache !== null &&
+    locationFinderCache[`${searchLocation.id}${maxDepth}`]
+  ) {
+    return locationFinderCache[`${searchLocation.id}${maxDepth}`];
```

Comment:
This may not need to be cached.  It should be one lookup value per group.  My reasoning was if things were recomputed we wouldn't have to search again. At that point the cache may be gone from re-rendering though.